File: Phasor.java
public class Phasor {
    public double rad;
    public double theta;

    public static final double pi = Math.PI;

    public static void main(String[] args) {
        Phasor i = p(20,210);
        int f = 5;

        Phasor z1 = add(parallel(l(4,f,0), r(20)),c(10,f,-3));
        Phasor temp1 = add(parallel(l(2,f,0), c(8,f,-3)),r(15));
        Phasor temp2 = add(l(8,f,0),r(40));
        Phasor z2 = add(temp1,temp2);

        System.out.println(idiv(z2,z1,i));
    }

    public static Phasor p(double rad, double theta) {
        return new Phasor(rad, theta);
    }

    public static Phasor r(double val) {
        return r(val,0);
    }

    public static Phasor r(double val, double exp) {
        return p(val * Math.pow(10,exp), 0);
    }

    public static Phasor c(double val, double freq, double exp) {
        return p(Math.pow(10,-exp) / (val * freq), -90);
    }

    public static Phasor l(double val, double freq, double exp) {
        return p(Math.pow(10,exp) * val * freq, 90);
    }

    public static Phasor p2(double a, double b) {
        double rad = Math.sqrt(a * a + b * b);
        double theta = Math.toDegrees(Math.atan2(b, a));
        return new Phasor(rad, theta);
    }

    public Phasor(double rad, double theta) {
        this.rad = rad;
        this.theta = theta;
    }

    public static Phasor add(Phasor p1, Phasor p2) {
        double x1 = p1.rad * Math.cos(Math.toRadians(p1.theta));
        double y1 = p1.rad * Math.sin(Math.toRadians(p1.theta));
        double x2 = p2.rad * Math.cos(Math.toRadians(p2.theta));
        double y2 = p2.rad * Math.sin(Math.toRadians(p2.theta));

        double x = x1 + x2;
        double y = y1 + y2;
        double r = Math.sqrt(x * x + y * y);
        double t = Math.toDegrees(Math.atan2(y, x));

        return new Phasor(r, t);
    }

    public static Phasor subtract(Phasor p1, Phasor p2) {
        double x1 = p1.rad * Math.cos(Math.toRadians(p1.theta));
        double y1 = p1.rad * Math.sin(Math.toRadians(p1.theta));
        double x2 = p2.rad * Math.cos(Math.toRadians(p2.theta));
        double y2 = p2.rad * Math.sin(Math.toRadians(p2.theta));

        double x = x1 - x2;
        double y = y1 - y2;
        double r = Math.sqrt(x * x + y * y);
        double t = Math.toDegrees(Math.atan2(y, x));

        return new Phasor(r, t);
    }

    public static Phasor multiply(Phasor p1, Phasor p2) {
        return new Phasor(p1.rad * p2.rad, p1.theta + p2.theta);
    }

    public static Phasor divide(Phasor p1, Phasor p2) {
        return new Phasor(p1.rad / p2.rad, p1.theta - p2.theta);
    }

    public Phasor inverse(){
        return divide(p(1,0),this);
    }
    public Phasor scale(int val){
        return p(rad*val,theta);
    }

    public Phasor negate(){
        return new Phasor(-rad,theta);
    }

    public static Phasor parallel(Phasor p1, Phasor p2) {
        return divide(new Phasor(1, 0), divide(add(p1, p2), multiply(p1, p2)));
    }

    public static Phasor series(Phasor p1, Phasor p2) {
        return add(p1, p2);
    }

    public static Phasor vdiv(Phasor z1, Phasor z2, Phasor v){
        return multiply(v, divide(z1, add(z1, z2)));
    }
    public static Phasor idiv(Phasor z1, Phasor z2, Phasor i){
        return multiply(i, divide(z2, add(z1, z2)));
    }

    public static Phasor[][] rref(Phasor[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int lead = 0;

        for (int r = 0; r < rows; r++) {
            if (lead >= cols) {
                return matrix;
            }
            int i = r;
            while (matrix[i][lead].rad == 0) {
                i++;
                if (i == rows) {
                    i = r;
                    lead++;
                    if (lead == cols) {
                        return matrix;
                    }
                }
            }
            Phasor[] temp = matrix[r];
            matrix[r] = matrix[i];
            matrix[i] = temp;

            Phasor lv = matrix[r][lead];
            for (int j = 0; j < cols; j++) {
                matrix[r][j] = divide(matrix[r][j], lv);
            }

            for (int i2 = 0; i2 < rows; i2++) {
                if (i2 != r) {
                    Phasor lv2 = matrix[i2][lead];
                    for (int j = 0; j < cols; j++) {
                        matrix[i2][j] = subtract(matrix[i2][j], multiply(lv2, matrix[r][j]));
                    }
                }
            }
            lead++;
        }
        return matrix;
    }

    @Override
    public String toString(){
        return Math.round(rad * 1000) / 1000.0 + " L " + Math.round(theta * 1000.0) / 1000.0;
    }

    public String toStringRect() {
        double a = rad * Math.cos(Math.toRadians(theta));
        double b = rad * Math.sin(Math.toRadians(theta));
        return Math.round(a * 1000) / 1000.0 + " + " + Math.round(b * 1000) / 1000.0 + "i";
    }
}

File: phasor.py
import math

class Phasor:
    """A class representing phasors (complex numbers in polar form)."""
    
    def __init__(self, magnitude: float, angle: float):
        self.magnitude = magnitude
        self.angle = angle  # In degrees

    def __repr__(self):
        return f"{round(self.magnitude, 3)} ∠ {round(self.angle, 3)}°"

    def to_rect(self):
        """Returns the rectangular form as (real, imaginary)."""
        real = self.magnitude * math.cos(math.radians(self.angle))
        imag = self.magnitude * math.sin(math.radians(self.angle))
        return round(real, 3), round(imag, 3)

    @classmethod
    def from_rect(cls, real: float, imag: float):
        """Creates a Phasor from rectangular (a + bi) form."""
        magnitude = math.sqrt(real**2 + imag**2)
        angle = math.degrees(math.atan2(imag, real))
        return cls(magnitude, angle)

    def p(magnitude: float, angle: float):
        """Creates a phasor in polar form (magnitude, angle)."""
        return Phasor(magnitude, angle)

    @classmethod
    def r(cls, val: float, exp: int = 0):
        """Creates a phasor representing a resistor."""
        return cls(val * (10 ** exp), 0)

    @classmethod
    def c(cls, val: float, freq: float, exp: int):
        """Creates a phasor for a capacitor."""
        return cls((10 ** -exp) / (val * freq), -90)

    @classmethod
    def l(cls, val: float, freq: float, exp: int):
        """Creates a phasor for an inductor."""
        return cls((10 ** exp) * val * freq, 90)

    def __add__(self, other):
        """Phasor addition in rectangular form."""
        real1, imag1 = self.to_rect()
        real2, imag2 = other.to_rect()
        return Phasor.from_rect(real1 + real2, imag1 + imag2)

    def __sub__(self, other):
        """Phasor subtraction in rectangular form."""
        real1, imag1 = self.to_rect()
        real2, imag2 = other.to_rect()
        return Phasor.from_rect(real1 - real2, imag1 - imag2)

    def __mul__(self, other):
        """Phasor multiplication (polar form)."""
        return Phasor(self.magnitude * other.magnitude, self.angle + other.angle)

    def __truediv__(self, other):
        """Phasor division (polar form)."""
        return Phasor(self.magnitude / other.magnitude, self.angle - other.angle)

    def inverse(self):
        """Returns the reciprocal of the phasor."""
        return Phasor(1 / self.magnitude, -self.angle)

    def scale(self, factor: float):
        """Scales the phasor magnitude by a factor."""
        return Phasor(self.magnitude * factor, self.angle)

    def negate(self):
        """Negates the phasor (180-degree phase shift)."""
        return Phasor(-self.magnitude, self.angle)

    @staticmethod
    def parallel(p1, p2):
        """Computes parallel impedance: (p1 * p2) / (p1 + p2)."""
        return (p1 * p2) / (p1 + p2)

    @staticmethod
    def series(p1, p2):
        """Computes series impedance: p1 + p2."""
        return p1 + p2

    @staticmethod
    def vdiv(z1, z2, v):
        """Voltage divider: V * (Z1 / (Z1 + Z2))"""
        return v * (z1 / (z1 + z2))

    @staticmethod
    def idiv(z1, z2, i):
        """Current divider: I * (Z2 / (Z1 + Z2))"""
        return i * (z2 / (z1 + z2))

p = Phasor.p
from_rect = Phasor.from_rect
r = Phasor.r
c = Phasor.c
l = Phasor.l
parallel = Phasor.parallel
series = Phasor.series
vdiv = Phasor.vdiv
idiv = Phasor.idiv
